File: prisma/refs/binary-tree-vertical-order-traversal.ts
type TreeNode = { val: number; left: TreeNode | null; right: TreeNode | null };

export function verticalOrder(root: TreeNode | null): number[][] {
  if (!root) return [];
  const cols = new Map<number, number[]>();
  const queue: [TreeNode, number][] = [[root, 0]];
  let min = 0, max = 0;
  while (queue.length) {
    const [node, c] = queue.shift()!;
    if (!cols.has(c)) cols.set(c, []);
    cols.get(c)!.push(node.val);
    min = Math.min(min, c);
    max = Math.max(max, c);
    if (node.left) queue.push([node.left, c - 1]);
    if (node.right) queue.push([node.right, c + 1]);
  }
  const out: number[][] = [];
  for (let c = min; c <= max; c++) out.push(cols.get(c) ?? []);
  return out;
}

File: prisma/refs/java/binary-tree-vertical-order-traversal.java
import java.util.*;

class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        List<List<Integer>> out = new ArrayList<>();
        if (root == null) return out;
        Map<Integer, List<Integer>> cols = new HashMap<>();
        Deque<TreeNode> nodes = new ArrayDeque<>();
        Deque<Integer> colq = new ArrayDeque<>();
        nodes.add(root);
        colq.add(0);
        int lo = 0, hi = 0;
        while (!nodes.isEmpty()) {
            TreeNode n = nodes.poll();
            int c = colq.poll();
            cols.computeIfAbsent(c, k -> new ArrayList<>()).add(n.val);
            lo = Math.min(lo, c);
            hi = Math.max(hi, c);
            if (n.left != null) { nodes.add(n.left); colq.add(c - 1); }
            if (n.right != null) { nodes.add(n.right); colq.add(c + 1); }
        }
        for (int c = lo; c <= hi; c++) out.add(cols.get(c));
        return out;
    }
}
